cpp/src/rolling/rolling.cu

```diff
@@ -649,12 +998,43 @@ std::unique_ptr<column> rolling_window(column_view const& input,
                                        size_type min_periods,
                                        std::unique_ptr<aggregation> const& agg,
                                        rmm::mr::device_memory_resource* mr)
+{
+  return rolling_window(
+    input, empty_like(input)->view(), preceding_window, following_window, min_periods, agg, mr);
+}
+
+// Applies a fixed-size rolling window function to the values in a column.
+std::unique_ptr<column> rolling_window(column_view const& input,
+                                       column_view const& default_outputs,
+                                       size_type preceding_window,
+                                       size_type following_window,
+                                       size_type min_periods,
+                                       std::unique_ptr<aggregation> const& agg,
+                                       rmm::mr::device_memory_resource* mr)
 {
   CUDF_FUNC_RANGE();
 
   if (input.size() == 0) return empty_like(input);
   CUDF_EXPECTS((min_periods >= 0), "min_periods must be non-negative");
 
+  CUDF_EXPECTS((default_outputs.is_empty() || default_outputs.size() == input.size()),
+               "Defaults column must be either empty or have as many rows as the input column.");
+
+  CUDF_EXPECTS(
+    default_outputs.is_empty() || agg->kind == aggregation::LEAD || agg->kind == aggregation::LAG,
+    "Only LEAD/LAG window functions support default values.");
+
+  CUDF_EXPECTS(default_outputs.type().id() == input.type().id(),
+               "Defaults column type must match input column.");  // Because LEAD/LAG.
+
+  if (agg->kind == aggregation::LEAD || agg->kind == aggregation::LAG) {
```

Comment:
All of the various `if(agg->kind...` checks should be eliminated. Agg specific logic should be in the agg specific code path. Otherwise we end up with a tangled mess of if/else logic checks.

I'd like to see this refactored to look like:
```
rolling_window(...){

// Do agg agnostic error checking
return agg_dispatcher(agg->kind, ...);

// all agg specific logic is put into the dispatched function object
}
```

Reply:
I wasn't fond of `aggr`-specific error checking either. Let me take a crack at using `agg_dispatcher` for this.

Reply:
I have moved `aggr`-specific code to the destination of the dispatch. Thank you, this is cleaner.